### PR TITLE
[BIOMAGE-1450] Enable navigation for processed project

### PR DESCRIPTION
**Pull request by StefanBabukov**
# Background
#### Link to issue 
https://biomage.atlassian.net/jira/software/c/projects/BIOMAGE/boards/1?modal=detail&selectedIssue=BIOMAGE-1450
#### Link to staging deployment URL 
https://ui-grasp6-ui572.scp-staging.biomage.net/data-management
#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Code updates
Have best practices and ongoing refactors being observed in this PR

- [ ] Migrated any selector / reducer used to the new format
- [ ] Validated that current unit tests work as expected and are enough

### Testing
- [ ] Unit tests written
- [ ] Tested locally with Inframock (with latest production data downloaded with biomage experiment pull)
- [ ] Deployed to staging
- [ ] Passed integration tests 

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/components/data-management/LaunchAnalysisButton.jsx
Comment (StefanBabukov) [inline src/components/data-management/LaunchAnalysisButton.jsx:59]: Adding this feature we want to make sure that the store is cleared when we navigate through the modules too, this is why this is removed and added in `ContentWrapper` under a condition so that we don't reset it on every browser refresh.

File: src/components/ContentWrapper.jsx
Comment (StefanBabukov) [inline src/components/ContentWrapper.jsx:45]: Changed this to differentiate between the experimentId in the URL and the one under activeproject.experiments[0]

File: src/__test__/components/ContentWrapper.test.jsx
Comment (aerlaut) [inline src/__test__/components/ContentWrapper.test.jsx:177]: Why is this divided by 2?

File: src/__test__/components/ContentWrapper.test.jsx
Comment (aerlaut) [inline src/__test__/components/ContentWrapper.test.jsx]: Can you use a loop to go through the menus instead of doing this manually?

File: src/__test__/components/ContentWrapper.test.jsx
Comment (aerlaut) [inline src/__test__/components/ContentWrapper.test.jsx:179]: Can we have a list containing the menu names, and then check for the length and the text of the menu too?

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [reply src/components/ContentWrapper.jsx:45]: When does this happen? The users can't change experiment if they are not in `data-management` right? and Data Management does not have a route in its URL

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: it seems `currentExpreiment` contains the `id`. Can you rename this variable to `currentExprimentId`? I think you should also change `currentExprerimentRef` to `currentExperimentIdRef`

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: What happens if we navigate to another experiment directly from the URL?

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: I think this is confusing. Instead of doing this, maybe you can do
```
const notProcessedExperimentDisable = routeExperimentId || pipelinesCompleted()
```

And then on line inside.`Menu.Item` `disabled` on line 323, you just do 
```
...
disabled={  notProcessedExperimentDisable || noExperimentDisable || pipelineStatusDisable }
...
```
With short-circuiting, this should be the same right?

File: src/__test__/components/ContentWrapper.test.jsx
Comment (StefanBabukov) [reply src/__test__/components/ContentWrapper.test.jsx:177]: Not sure, each link has 2 items  looking at the DOM they are different, here are 2 trees for the same link: 
`  <MenuItem id="/experiments/[experimentId]/plots-and-tables" disabled={false} icon={{...}} onClick={[Function: onClick]} eventKey="/experiments/[experimentId]/plots-and-tables">
      <Tooltip title={{...}} visible={false} placement="right" overlayClassName="ant-menu-inline-collapsed-tooltip" mouseEnterDelay={0.1} mouseLeaveDelay={0.1} arrowPointAtCenter={false} autoAdjustOverflow={true}>
        <ForwardRef(Tooltip) title={{...}} visible={false} placement="right" overlayClassName="ant-menu-inline-collapsed-tooltip" mouseEnterDelay={0.1} mouseLeaveDelay={0.1} arrowPointAtCenter={false} autoAdjustOverflow={true} prefixCls="ant-tooltip" getTooltipContainer={[undefined]} builtinPlacements={{...}} overlay="" onVisibleChange={[Function: onVisibleChange]} onPopupAlign={[Function: onPopupAlign]} overlayInnerStyle={[undefined]} arrowContent={{...}} motion={{...}}>
          <Trigger popupClassName="ant-menu-inline-collapsed-tooltip" prefixCls="ant-tooltip" popup={[Function: getPopupElement]} action={{...}} builtinPlacements={{...}} popupPlacement="right" popupAlign={{...}} getPopupContainer={[undefined]} onPopupVisibleChange={[Function: onVisibleChange]} afterPopupVisibleChange={[Function: noop]} popupTransitionName={[undefined]} popupAnimation={[undefined]} popupMotion={{...}} defaultPopupVisible={false} destroyPopupOnHide={false} autoDestroy={false} mouseLeaveDelay={0.1} popupStyle={{...}} mouseEnterDelay={0.1} title={{...}} visible={false} arrowPointAtCenter={false} autoAdjustOverflow={true} overlay="" onPopupAlign={[Function: onPopupAlign]} arrowContent={{...}} popupVisible={false} getPopupClassNameFromAlign={[Function: returnEmptyString]} getDocument={[Function: returnDocument]} focusDelay={0} blurDelay={0.15} mask={false} maskClosable={true} showAction={{...}} hideAction={{...}}>
            <MenuItem id="/experiments/[experimentId]/plots-and-tables" disabled={false} onClick={[Function: onClick]} eventKey="/experiments/[experimentId]/plots-and-tables" className="" title={[undefined]} onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onFocus={[undefined]} onBlur={[undefined]}>
              <InternalMenuItem id="/experiments/[experimentId]/plots-and-tables" disabled={false} onClick={[Function: onClick]} eventKey="/experiments/[experimentId]/plots-and-tables" className="" title={[undefined]} onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onFocus={[undefined]} onBlur={[undefined]}>
                <LegacyMenuItem elementRef={{...}} role="menuitem" tabIndex={-1} data-menu-id="rc-menu-uuid-test-/experiments/[experimentId]/plots-and-tables" id="/experiments/[experimentId]/plots-and-tables" title={[undefined]} onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onBlur={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} component="li" aria-disabled={false} style={{...}} className="ant-menu-item" onClick={[Function: onInternalClick]} onKeyDown={[Function: onInternalKeyDown]} onFocus={[Function: onInternalFocus]}>
                  <RawItem title={[undefined]} role="menuitem" tabIndex={-1} data-menu-id="rc-menu-uuid-test-/experiments/[experimentId]/plots-and-tables" id="/experiments/[experimentId]/plots-and-tables" onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onBlur={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} component="li" aria-disabled={false} style={{...}} className="ant-menu-item" onClick={[Function: onInternalClick]} onKeyDown={[Function: onInternalKeyDown]} onFocus={[Function: onInternalFocus]}>
                    <Item className="ant-menu-item" prefixCls="ant-menu-overflow-item" responsive={false} component="li" invalidate={true} order={1} item={{...}} itemKey={1} registerSize={[Function: registerSize]} display={true} title={[undefined]} role="menuitem" tabIndex={-1} data-menu-id="rc-menu-uuid-test-/experiments/[experimentId]/plots-and-tables" id="/experiments/[experimentId]/plots-and-tables" onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onBlur={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} aria-disabled={false} style={{...}} onClick={[Function: onInternalClick]} onKeyDown={[Function: onInternalKeyDown]} onFocus={[Function: onInternalFocus]}>
                      <li className="ant-menu-item" style={{...}} title={[undefined]} role="menuitem" tabIndex={-1} data-menu-id="rc-menu-uuid-test-/experiments/[experimentId]/plots-and-tables" id="/experiments/[experimentId]/plots-and-tables" onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onBlur={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} aria-disabled={false} onClick={[Function: onInternalClick]} onKeyDown={[Function: onInternalKeyDown]} onFocus={[Function: onInternalFocus]}>
                        <ForwardRef(DatabaseOutlined) className="ant-menu-item-icon">
                          <AntdIcon className="ant-menu-item-icon" icon={{...}}>
                            <span role="img" aria-label="database" tabIndex={[undefined]} onClick={[undefined]} className="anticon anticon-database ant-menu-item-icon">
                              <IconReact icon={{...}} primaryColor={[undefined]} secondaryColor={[undefined]} style={[undefined]}>
                                <svg viewBox="64 64 896 896" focusable="false" className={[undefined]} onClick={[undefined]} style={[undefined]} data-icon="database" width="1em" height="1em" fill="currentColor" aria-hidden="true">
                                  <path d="M832 64H192c-17.7 0-32 14.3-32 32v832c0 17.7 14.3 32 32 32h640c17.7 0 32-14.3 32-32V96c0-17.7-14.3-32-32-32zm-600 72h560v208H232V136zm560 480H232V408h560v208zm0 272H232V680h560v208zM304 240a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0zm0 272a40 40 0 1080 0 40 40 0 10-80 0z" />
                                </svg>
                              </IconReact>
                            </span>
                          </AntdIcon>
                        </ForwardRef(DatabaseOutlined)>
                        <span className="ant-menu-title-content">
                          Plots and Tables
                        </span>
                        <Icon props={{...}} icon={[undefined]} />
                      </li>
                    </Item>
                  </RawItem>
                </LegacyMenuItem>
              </InternalMenuItem>
            </MenuItem>
          </Trigger>
        </ForwardRef(Tooltip)>
      </Tooltip>
    </MenuItem>`

`    <MenuItem id="/experiments/[experimentId]/plots-and-tables" disabled={false} icon={{...}} onClick={[Function: onClick]} eventKey="/experiments/[experimentId]/plots-and-tables">
      <Tooltip title={{...}} visible={false} placement="right" overlayClassName="ant-menu-inline-collapsed-tooltip" mouseEnterDelay={0.1} mouseLeaveDelay={0.1} arrowPointAtCenter={false} autoAdjustOverflow={true}>
        <ForwardRef(Tooltip) title={{...}} visible={false} placement="right" overlayClassName="ant-menu-inline-collapsed-tooltip" mouseEnterDelay={0.1} mouseLeaveDelay={0.1} arrowPointAtCenter={false} autoAdjustOverflow={true} prefixCls="ant-tooltip" getTooltipContainer={[undefined]} builtinPlacements={{...}} overlay="" onVisibleChange={[Function: onVisibleChange]} onPopupAlign={[Function: onPopupAlign]} overlayInnerStyle={[undefined]} arrowContent={{...}} motion={{...}}>
          <Trigger popupClassName="ant-menu-inline-collapsed-tooltip" prefixCls="ant-tooltip" popup={[Function: getPopupElement]} action={{...}} builtinPlacements={{...}} popupPlacement="right" popupAlign={{...}} getPopupContainer={[undefined]} onPopupVisibleChange={[Function: onVisibleChange]} afterPopupVisibleChange={[Function: noop]} popupTransitionName={[undefined]} popupAnimation={[undefined]} popupMotion={{...}} defaultPopupVisible={false} destroyPopupOnHide={false} autoDestroy={false} mouseLeaveDelay={0.1} popupStyle={{...}} mouseEnterDelay={0.1} title={{...}} visible={false} arrowPointAtCenter={false} autoAdjustOverflow={true} overlay="" onPopupAlign={[Function: onPopupAlign]} arrowContent={{...}} popupVisible={false} getPopupClassNameFromAlign={[Function: returnEmptyString]} getDocument={[Function: returnDocument]} focusDelay={0} blurDelay={0.15} mask={false} maskClosable={true} showAction={{...}} hideAction={{...}}>
            <MenuItem id="/experiments/[experimentId]/plots-and-tables" disabled={false} onClick={[Function: onClick]} eventKey="/experiments/[experimentId]/plots-and-tables" className="" title={[undefined]} onContextMenu={[undefined]} onMouseDown={[undefined]} onTouchStart={[undefined]} onMouseEnter={[Function (anonymous)]} onMouseLeave={[Function (anonymous)]} onFocus={[undefined]} onBlur={[undefined]} />
          </Trigger>
        </ForwardRef(Tooltip)>
      </Tooltip>
    </MenuItem>`

File: src/components/ContentWrapper.jsx
Comment (StefanBabukov) [reply src/components/ContentWrapper.jsx]: If we are not in data-management it won't get reset, so we reset it only if the user is in data-management that way we know they are launching a project.

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: You shouldn't need the parantheses no?

**Comment by aerlaut** [comment]
Can you also enable the experiment name on the left side of the bar?

**Comment by vickymorrison** [comment]
In the admin account I found a project named 'test mh ui' where the Go to Data Processing button is disabled with a message to ensure all data is uploaded and metadata is input. However, the navigation controls for the other modules are enabled on the left hand side. (See screenshot.) This should not happen - either both should be disabled or both should be enabled.
Screenshot 2021-12-02 at 20 16 34

**Comment by vickymorrison** [comment]
I also found one that is the other way about - project 'Test of Stefans PR' has the Go to Data Processing button enabled and yet the navigation controls on the left side are disabled. Again, this should not happen.
Screenshot 2021-12-02 at 20 20 11

**Comment by StefanBabukov** [comment]
@vickymorrison  I believe the last screenshot is due to that the QC is not succeeded yet and this is why the left links are disabled

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (aerlaut) [inline src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Can you use a more readable mock experiment name than this?

File: src/__test__/utils/AppRouteProvider.test.jsx
Comment (aerlaut) [inline src/__test__/utils/AppRouteProvider.test.jsx]: Can you use a more readable mock projectUuid than this?

File: src/__test__/utils/AppRouteProvider.test.jsx
Comment (aerlaut) [inline src/__test__/utils/AppRouteProvider.test.jsx]: Can you use a more readable mock paramsHash than this?

File: src/__test__/utils/AppRouteProvider.test.jsx
Comment (aerlaut) [inline src/__test__/utils/AppRouteProvider.test.jsx]: Can you put a space between one test case test and the next?

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: Can you put a space before and after `notProcessedExperimentDisable` to make it easier to read?

**Comment by aerlaut** [comment]
I'd change the wordings to:

**You have made changes that require data to be processed again.**
Would you like to discard these changes and navigate to the page using old data or re-run sample processing?

and the buttons :
Discard changes and continue 
Re-run sample processing

what do you think @vickymorrison?

image

File: src/components/data-management/DataManagementIntercept.jsx
Comment (aerlaut) [inline src/components/data-management/DataManagementIntercept.jsx]: Clicking on the `x` close button on the top right part of the modal doesn't dismiss it. Add a handler for that event by adding a callback to `onCancel` on the modal.

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (aerlaut) [inline src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Add tests for : 
- Discarding the changes and continuing to the path
- Dismissing the modal using the close button (the x on the top right)

**Comment by aerlaut** [comment]
I see that the navigation menus are disabled when QC pipeline fails. Can you add tooltips to the navigation menu links which say "Data Processing failed. Click Go to Data Processing to continue." to let the user know they have to go to Data Processing and re-run the pipeline?

https://user-images.githubusercontent.com/2862362/145179945-d2c6f1aa-6ea1-470d-8db2-70f7cccfb4f5.mov

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (StefanBabukov) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: There is already a test - Dispatches routes correctly for continuing the route in AppRouteProvider.test, which is basically testing the onContinue function that is passed to DataManagementIntercept as well. `onDismissIntercept` is also tested, which is used for `onCancel`.

Comment (StefanBabukov) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Or am i missing something?

Comment (aerlaut) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: The one that we want to test here is the interface, the the buttons in `DataManagementIntercept` dispatch the correct actions. You have tested one of them, the re-process button, but there are 2 more actions a user can take, clicking on the discard button and clicking on the close button. So those are the 2 interactions that we want to test for,

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (aerlaut) [inline src/__test__/components/data-management/DataManagementIntercept.test.jsx]: I think it's better to move this test into it's own case, to test that clicking the button triggers `onContinue`

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (aerlaut) [inline src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Maybe rename this test case to something like "Clicking re-process button re-runs GEM2S"

File: src/__test__/components/data-management/DataManagementIntercept.test.jsx
Comment (aerlaut) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Oh I see that you have tested for clicking the `continue` in the first case. Can you split the test case to make it clearer?

Comment (aerlaut) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Then there is only 1 test case left to handle, which is clicking the `x` button on the top right part of the modal.

Comment (aerlaut) [reply src/__test__/components/data-management/DataManagementIntercept.test.jsx]: Aside from checking `onDismissIntercept` is triggered, is it possible to check if the element does not exist?

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: if the experiment is not processed, and we are in `data-management`, don't we also not have experiment id?

Comment (StefanBabukov) [reply src/components/ContentWrapper.jsx]: no, thats the experimentId in the URL that we don't have, this is why I get the one under the activeProject

Comment (aerlaut) [reply src/components/ContentWrapper.jsx]: Ooh I see.. Is this comment is about the `useEffect` block below? Can you change this comment to make it clearer? Or maybe delete it if we don't need it,

File: src/components/ContentWrapper.jsx
Comment (aerlaut) [inline src/components/ContentWrapper.jsx]: Does moving this code below the `useEffect` affect the code? can you move below the block? I think it's clearer if it's placed there, because then you know what the value of `currentExperimentIdRef`resolves to.

**Review by aerlaut — APPROVED** [review]
code LGTM. Please get @vickymorrison's approval before merging,

**Review by vickymorrison — APPROVED** [review]

**Pull request merged**
